plugins/c-ares-dns-client.c

```diff
@@ -0,0 +1,757 @@
+	if (request->ares_watches == NULL) {
+		DBG("could not create ares_watches table");
+		g_free(request);
+		return NULL;
+	}
+
+	/*
+	 * Init a c-ares channel for this request. We set an option asking
+	 * c-ares to notify us via callback about which sockets it wants to
+	 * monitor for readability and writability. This allows us to
+	 * integrate c-ares activity into our glib main event loop.
+	 */
+	memset(&options, 0, sizeof(options));
+	options.sock_state_cb = ares_socket_state_cb;
+	options.sock_state_cb_data = request;
+	optmask = ARES_OPT_SOCK_STATE_CB;
+	if (timeout_ms > 0) {
+		options.timeout = timeout_ms;
+		optmask |= ARES_OPT_TIMEOUTMS;
+	}
+	ares_status = ares_init_options(&request->channel, &options, optmask);
+	if (ares_status != ARES_SUCCESS) {
+		ofono_error("%s: failed to init c-ares channel: %s", __func__,
+				ares_strerror(ares_status));
+		request->running = FALSE;  /* don't trip assertion */
+		g_hash_table_destroy(request->ares_watches);
+		g_free(request);
+		return NULL;
+	}
+
+	if (servers != NULL) {
+		set_request_servers(servers, request);
+		ares_status =
+			ares_set_servers(request->channel, request->servers);
+		/* Log an error, but continue anyway */
+		if (ares_status != ARES_SUCCESS)
+			ofono_error("%s: failed to set name servers: %s",
+					__func__, ares_strerror(ares_status));
+	}
+
+	/*
+	 * If the caller has provided a preferred interface, tell c-ares to
+	 * send requests out that interface, and unset rf filter.
+	 */
+	if (interface != NULL) {
+		DBG("caller has specified device %s", interface);
+
+		request->interface = g_strdup(interface);
+		ares_set_local_dev(request->channel, request->interface);
+		rp_filter_set(request->interface, FALSE);
+	}
+
+	request->cb = cb;
+	request->data = data;
+	request->hostname = g_strdup(hostname);
+	request->timeout.tv_sec = timeout_ms / 1000;
+	request->timeout.tv_usec = (timeout_ms % 1000) * 1000;
+	gettimeofday(&request->start_time, NULL);
+
+	pending_requests = g_list_append(pending_requests, request);
+
+	ares_gethostbyname(request->channel, hostname, AF_INET,
+						ares_request_cb, request);
+
+	reset_ares_timeout(request, destroy_old_source);
+
+	return request;
+}
+
+/* Cancel an in-progress name resolution request. */
+static gboolean ofono_dns_client_cancel_request(ofono_dns_client_request_t req)
+{
+	struct ares_request *request = req;
+
+	if (request == NULL || g_list_find (pending_requests, request) == NULL)
+		return FALSE;
+
+	DBG("request %p", request);
+
+	if (!request->running)
+		return TRUE;
+
+	pending_requests = g_list_remove(pending_requests, request);
+	request->running = FALSE;  /* don't trip assertion */
+	destroy_ares_request(request);
+
+	return TRUE;
+}
+
+static struct ofono_dns_client_driver dns_driver = {
+	.name		= "c-ares DNS client",
+	.submit_request	= ofono_dns_client_submit_request,
+	.cancel_request = ofono_dns_client_cancel_request
+};
+
+/* Intitialize this module. */
+static int c_ares_init(void)
+{
```

Comment:
I'm assuming the init() and exit() functions are new?  If so, I think it'd be appropriate to follow the ofono ws rules ( blank lines before/after if blocks )... If not, please disregard this comment.